File: js/src/absolute-path.js
// @flow

// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

import {invariant, notNull} from './assert.js';
import {datasetRe} from './dataset.js';
import Database from './database.js';
import {default as Hash, stringLength} from './hash.js';
import Path from './path.js';
import type Value from './value.js';

const datasetCapturePrefixRe = new RegExp('^(' + datasetRe.source + ')');

/**
 * An AbsolutePath is a Path relative to either a dataset head, or a hash.
 */
export default class AbsolutePath {
  /** The dataset ID that `_path` is in, or `''` if none. */
  dataset: string;

  /** The hash the that `_path` is in, if any. */
  hash: Hash | null;

  /** Path relative to either `_dataset` or `_hash`. */
  path: Path;

  /**
   * Returns `str` parsed as an AbsolutePath if successful, or a null path with an error message
   * if not.
   */
  static parse(str: string): [AbsolutePath | null, string] {
    if (str === '') {
      return [null, 'Empty path'];
    }

    let dataset = '';
    let hash = null;
    let pathStr = '';

    if (str[0] === '#') {
      const tail = str.slice(1);
      if (tail.length < stringLength) {
        return [null, `Invalid hash: ${tail}`];
      }

      const hashStr = tail.slice(0, stringLength);
      hash = Hash.parse(hashStr);
      if (hash === null) {
        return [null, `Invalid hash: ${hashStr}`];
      }

      pathStr = tail.slice(stringLength);
    } else {
      const parts = datasetCapturePrefixRe.exec(str);
      if (!parts) {
        return [null, `Invalid dataset name: ${str}`];
      }

      invariant(parts.length === 2);
      dataset = parts[1];
      pathStr = str.slice(parts[0].length);
    }

    if (pathStr.length === 0) {
      return [new AbsolutePath(dataset, hash, new Path()), ''];
    }

    const [path, err] = Path.parse(pathStr);
    if (err !== '') {
      return [null, err];
    }

    return [new AbsolutePath(dataset, hash, notNull(path)), ''];
  }

  constructor(dataset: string, hash: Hash | null, path: Path) {
    this.dataset = dataset;
    this.hash = hash;
    this.path = path;
  }

  async resolve(db: Database): Promise<Value | null> {
    let val = null;
    if (this.dataset !== '') {
      val = await db.head(this.dataset);
    } else if (this.hash !== null) {
      val = await db.readValue(this.hash);
    } else {
      throw new Error('unreachable');
    }

    if (val === undefined) {
      val = null;
    }

    if (val !== null) {
      return this.path.resolve(val);
    } else {
      return null;
    }
  }

  toString(): string {
    if (this.dataset !== '') {
      return this.dataset + this.path.toString();
    }
    if (this.hash !== null) {
      return '#' + this.hash.toString() + this.path.toString();
    }
    throw new Error('unreachable');
  }
}

File: go/spec/absolute_path.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package spec

import (
	"errors"
	"fmt"
	"regexp"

	"github.com/attic-labs/noms/go/d"
	"github.com/attic-labs/noms/go/datas"
	"github.com/attic-labs/noms/go/hash"
	"github.com/attic-labs/noms/go/types"
)

var datasetCapturePrefixRe = regexp.MustCompile("^(" + datas.DatasetRe.String() + ")")

type AbsolutePath struct {
	Dataset string
	Hash    hash.Hash
	Path    types.Path
}

func NewAbsolutePath(str string) (AbsolutePath, error) {
	if len(str) == 0 {
		return AbsolutePath{}, errors.New("Empty path")
	}

	var h hash.Hash
	var dataset string
	var pathStr string

	if str[0] == '#' {
		tail := str[1:]
		if len(tail) < hash.StringLen {
			return AbsolutePath{}, errors.New("Invalid hash: " + tail)
		}

		hashStr := tail[:hash.StringLen]
		if h2, ok := hash.MaybeParse(hashStr); ok {
			h = h2
		} else {
			return AbsolutePath{}, errors.New("Invalid hash: " + hashStr)
		}

		pathStr = tail[hash.StringLen:]
	} else {
		datasetParts := datasetCapturePrefixRe.FindStringSubmatch(str)
		if datasetParts == nil {
			return AbsolutePath{}, fmt.Errorf("Invalid dataset name: %s", str)
		}

		dataset = datasetParts[1]
		pathStr = str[len(dataset):]
	}

	if len(pathStr) == 0 {
		return AbsolutePath{Hash: h, Dataset: dataset}, nil
	}

	path, err := types.ParsePath(pathStr)
	if err != nil {
		return AbsolutePath{}, err
	}

	return AbsolutePath{Hash: h, Dataset: dataset, Path: path}, nil
}

func (p AbsolutePath) Resolve(db datas.Database) (val types.Value) {
	if len(p.Dataset) > 0 {
		var ok bool
		ds := db.GetDataset(p.Dataset)
		if val, ok = ds.MaybeHead(); !ok {
			val = nil
		}
	} else if !p.Hash.IsEmpty() {
		val = db.ReadValue(p.Hash)
	} else {
		d.Chk.Fail("Unreachable")
	}

	if val != nil && p.Path != nil {
		val = p.Path.Resolve(val)
	}
	return
}

func (p AbsolutePath) String() (str string) {
	if len(p.Dataset) > 0 {
		str = p.Dataset
	} else if !p.Hash.IsEmpty() {
		str = "#" + p.Hash.String()
	} else {
		d.Chk.Fail("Unreachable")
	}

	return str + p.Path.String()
}

func ReadAbsolutePaths(db datas.Database, paths ...string) ([]types.Value, error) {
	r := make([]types.Value, 0, len(paths))
	for _, ps := range paths {
		p, err := NewAbsolutePath(ps)
		if err != nil {
			return nil, fmt.Errorf("Invalid input path '%s'", ps)
		}

		v := p.Resolve(db)
		if v == nil {
			return nil, fmt.Errorf("Input path '%s' does not exist in database", ps)
		}

		r = append(r, v)
	}
	return r, nil
}
